controllers/capabilities/backend_controller.go

```diff
@@ -240,71 +238,95 @@ func (r *BackendReconciler) validateSpec(backendResource *capabilitiesv1beta1.Ba
 	}
 }
 
-func (r *BackendReconciler) removeBackend(providerAccountRef *corev1.LocalObjectReference, backendID int64, backendNamespace string, systemName string) (requeue bool, err error) {
-	providerAccount, err := controllerhelper.LookupProviderAccount(r.Client(), backendNamespace, providerAccountRef, r.Logger())
-	if err != nil {
-		return false, err
-	}
-
-	threescaleAPIClient, err := controllerhelper.PortaClient(providerAccount)
-	if err != nil {
-		return false, err
-	}
-
+func (r *BackendReconciler) removeBackendReferencesFromProducts(backend *capabilitiesv1beta1.Backend) (ctrl.Result, error) {
 	// Retrieve all product CRs that are under the same ns as the backend CR
 	opts := k8sclient.ListOptions{
-		Namespace: backendNamespace,
+		Namespace: backend.Namespace,
 	}
 	productCRsList := &capabilitiesv1beta1.ProductList{}
-	err = r.Client().List(context.TODO(), productCRsList, &opts)
+	err := r.Client().List(context.TODO(), productCRsList, &opts)
 	if err != nil {
-		return false, err
+		return ctrl.Result{}, err
 	}
 
 	// fetch CRs that belong to a tenant and require removal of the backend mentions in
 	// Application Plan pricing rules
 	// Application Plan limits
 	// Backend usages
-	tenantProductCRs, err := r.fetchTenantProductCRs(productCRsList, providerAccountRef, backendNamespace, systemName)
+	tenantProductCRs, err := r.fetchTenantProductCRs(productCRsList, backend)
 	if err != nil {
-		return false, err
+		return ctrl.Result{}, err
 	}
 
-	var productCrUpdated = false
+	productCRUpdated := false
 
 	// update backendUsages for each product retrieved
 	for _, productCR := range tenantProductCRs {
-		if productCR.RemoveBackendReferences(systemName) {
+		if productCR.RemoveBackendReferences(backend.Spec.SystemName) {
 			err = r.UpdateResource(&productCR)
 			if err != nil {
-				return false, err
+				return ctrl.Result{}, err
 			}
-			productCrUpdated = true
+			productCRUpdated = true
 		}
 	}
 
-	if productCrUpdated {
-		return true, nil
+	if productCRUpdated {
+		return ctrl.Result{Requeue: productCRUpdated, RequeueAfter: requeueTime}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *BackendReconciler) removeBackendFrom3scale(backend *capabilitiesv1beta1.Backend) error {
+	logger := r.Logger().WithValues("backend", client.ObjectKey{Name: backend.Name, Namespace: backend.Namespace})
+
+	// Attempt to remove backend only if backend.Status.ID is present
+	if backend.Status.ID == nil {
+		logger.Info("could not remove backend because backend ID is missing for backend name")
```

Comment:
if the ID is nil I think 305 will be throwing nil pointer?

Reply:
good catch! I forgot the early `return nil`